src/server/scene/application_session.cpp

```diff
@@ -81,9 +81,15 @@ ms::ApplicationSession::~ApplicationSession()
 }
 
 auto ms::ApplicationSession::create_surface(
+    std::shared_ptr<Session> const& session,
     SurfaceCreationParameters const& the_params,
     std::shared_ptr<ms::SurfaceObserver> const& observer) -> std::shared_ptr<Surface>
 {
+    if (session && session.get() != this)
+    {
+        BOOST_THROW_EXCEPTION(std::logic_error("Incorrect session"));
```

Comment:
I realize this is a precondition failure, but I don't think we should expect execution to continue. I would prefer `fatal_error()`

Reply:
I disagree. If we have a bug that only triggers in specific conditions (perhaps it only happens on clients using a specific protocol), we should __not__ crash the whole server over that. Disconnecting one client (as I assume throwing an exception will) should be disruptive enough to get our attention.

Reply:
AFAICS the only reason for passing `session` to `create_surface()` is that we can't create a `shared_ptr` from `this`?

I've spent too much time trying to figure out the original cause of problems in systems that "carried on regardless" (or relied on someone spotting a log message to have confidence that anything less than a core will be "enough to get our attention". We ought to be covering any invocations of this function in our testing: the very few call points for this function and they should all be under our control.

Your scenario that some client using a specific protocol finds a codepath we haven't covered in our tests is unlikely but, if it does happen, we'll likely never hear about it unless we do more than hope we disconnect the client.

Reply:
Fair enough